src/community/ncwms/src/main/java/org/geoserver/wms/ncwms/NcWmsService.java

```diff
@@ -70,6 +75,58 @@ public void checkTimeout() {
         }
     }
 
+    enum DateFinder {
+        NEAREST {
+            // This is used when nearest match is supported.
+            @Override
+            List<DateRange> findDates(WMS wms, CoverageInfo coverage, List<Object> times)
+                    throws IOException {
+                TreeSet<Date> availableDates = wms.queryCoverageNearestMatchTimes(coverage, times);
+                return availableDates
+                        .stream()
+                        .map(date -> new DateRange(date, date))
+                        .collect(Collectors.toList());
+            }
+        },
+        QUERY {
+            // When nearest match isn't enabled, let's query the coverage to identify
+            // available dates only
+            @Override
+            List<DateRange> findDates(WMS wms, CoverageInfo coverage, List<Object> times)
+                    throws IOException {
+                TreeSet<Date> availableDates = new TreeSet<Date>();
+                final boolean isRange = times.get(0) instanceof DateRange;
+                for (Object time : times) {
+                    TreeSet<Object> foundTimes =
+                            wms.queryCoverageTimes(
+                                    coverage, getAsRange(time, isRange), Query.DEFAULT_MAX);
+                    foundTimes
+                            .stream()
+                            .sequential()
+                            .forEach(
+                                    d -> {
+                                        Date date = (Date) d;
+                                        if (!availableDates.contains(date)) {
```

Comment:
It's a set, no need for explicit checks, just add

Reply:
Right.